### PR TITLE
Return the fileid from `copyFromCache` and use it instead of doing an extra query

### PR DESCRIPTION
**Pull request by icewind1991**
Also move the existing `copyFromCache` code around and expose it trough the public interface.

### CODE REVIEW

**Review by rullzer — APPROVED** [review]
Didn't test but the code makes sense!

File: lib/private/Files/Cache/Cache.php [hunk 1]
```diff
@@ -1076,4 +1076,44 @@ public static function getById($id) {
...
+		$data = $this->cacheEntryToArray($sourceEntry);
+		$fileId = $this->put($targetPath, $data);
+		if ($fileId <= 0) {
+			throw new \RuntimeException("Failed to copy to " . $targetPath . " from cache with source data " . json_encode($data) . " ");
+		}
+		if ($sourceEntry->getMimeType() === ICacheEntry::DIRECTORY_MIMETYPE) {
+			$folderContent = $sourceCache->getFolderContentsById($sourceEntry->getId());
+			foreach ($folderContent as $subEntry) {
+				$subTargetPath = $targetPath . '/' . $subEntry->getName();
+				$this->copyFromCache($sourceCache, $subEntry, $subTargetPath);
```
Comment (PVince81) [inline lib/private/Files/Cache/Cache.php:1098]: I guess we're not afraid of very deep folder structures and stack overflows ?

Edit: just saw it's old copied code, but something to keep in mind as tech debt?

File: lib/private/Files/Cache/Cache.php [hunk 2]
```diff
@@ -1076,4 +1076,44 @@ public static function getById($id) {
...
+	/**
+	 * Copy a file or folder in the cache
+	 *
+	 * @param ICache $sourceCache
+	 * @param ICacheEntry $sourceEntry
+	 * @param string $targetPath
+	 * @return int fileid of copied entry
+	 */
+	public function copyFromCache(ICache $sourceCache, ICacheEntry $sourceEntry, string $targetPath): int {
+		$data = $this->cacheEntryToArray($sourceEntry);
```
Comment (PVince81) [inline lib/private/Files/Cache/Cache.php:1089]: I guess $sourceCache will never be a FailedCache here ?

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 3]
```diff
@@ -567,14 +568,13 @@ private function copyFile(ICacheEntry $sourceEntry, string $to) {
 
 		$sourceUrn = $this->getURN($sourceEntry->getId());
 
-		$cache->copyFromCache($cache, $sourceEntry, $to);
-		$targetEntry = $cache->get($to);
-
-		if (!$targetEntry) {
-			throw new \Exception('Target not in cache after copy');
+		if (!$cache instanceof Cache) {
```
Comment (PVince81) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:571]: check for FailedCache ?

**Review by PVince81 — APPROVED** [review]
:+1: Looks fine, see comments if applicable

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php (hunk 3 above)
Comment (icewind1991) [reply lib/private/Files/ObjectStore/ObjectStoreStorage.php:571]: `FailedCache` isn't an `OC\Filesystem\Cache\Cache` only an `OCP\Filesystem\Cache\ICache`

File: lib/private/Files/Cache/Cache.php (hunk 2 above)
Comment (icewind1991) [reply lib/private/Files/Cache/Cache.php:1089]: I think so yes, added a safety check anyway

**Comment by PVince81** [comment]
build restarted...

**Pull request merged**